Section on reading from blobs with Python

CPBridge opened the pull request:
@fedorov here is a new page on using python tools to read directly from blob objects

This is the first time I've used gitbook, so not sure if there's anything else I need to do to make sure the page is included etc?

CPBridge committed:
Add page on direct loading of data

CPBridge committed:
Various tweaks

CPBridge committed:
Adapt image size

CPBridge committed:
Adjust image again

CPBridge committed:
Center the image

CPBridge committed:
Correct image size again

CPBridge committed:
Further minor tweaks

CPBridge committed:
Add page to summar.md

CPBridge committed:
Add section on offset tables

CPBridge committed:
Fix typo, minor clarification

CPBridge commented:
Ok I figured it out, it appears to be showing correctly in the preview

CPBridge commented:
I'm getting a load of errors in the CI pipeline about incorrect URLs, but these seem unrelated to my changes

CPBridge committed:
Further typo

CPBridge committed:
remove extra #

fedorov commented:
@CPBridge thank you for the PR! I would like to first fix https://github.com/ImagingDataCommons/idc-index/issues/102, since otherwise it will be difficult to explain where users can get those file paths.

Sorry about the false alarms with the failing checks.

CPBridge committed:
Add S3 examples

CPBridge committed:
Change title to match conventio

CPBridge committed:
Re-organization of pydicom section

CPBridge committed:
Fix text wrapping

CPBridge committed:
Update index entry title

CPBridge committed:
Rephrase pydicom section

CPBridge committed:
Further restructuring

CPBridge committed:
GDS -> GCS

CPBridge committed:
title capitalisation

CPBridge committed:
GCS typo

fedorov commented:
@CPBridge I finally implemented support for getting URLs from GCS, and while testing it, I discovered that it seems that the current sample code at least requires the user to authenticate and provide credentials. Is this right? Do you know if it is possible to somehow access the bucket without signing in - since the content is public. idc-index can download the entire file without login.

You can see the error and the updated part getting file URLs via idc-index in this notebook: https://colab.research.google.com/drive/1_5cc_kb-FgUl9r0jePX_QbSvwBFEluMg?usp=sharing

CPBridge committed:
Switch to anonymous client

CPBridge commented:
@fedorov this can be avoided by using an anonymous client. I have updated the examples in the PR to demonstrate this

fedorov commented:
@CPBridge FYI, the below takes ~30 sec on Colab (and about 47 sec when I ran it first). Is this expected?

![image](https://github.com/user-attachments/assets/d8a993a6-da5e-4666-a069-af71b6e146e3)

CPBridge committed:
Adapt S3 reading recommendations

CPBridge committed:
fix line wrapping

CPBridge commented:
@fedorov I have updated the recommendations for S3 in most cases to avoid the slow behavior you observed.

I tracked it down to the way smart_open buffers data. It maintains a buffer of data it pulls down but it re-populates this whenever you do a `seek` operation, even if the new seek position exists within the current buffer. This is problematic because pydicom does a lot of little seek operations when reading in order to jump around through tags. There is some discussion on this topic on the smart_open repo [here](https://github.com/piskvorky/smart_open/issues/712). I just made some suggestions, we'll see if they are interested in improving this.

CPBridge committed:
Use context managers for open methods

CPBridge committed:
Fix minor typo

CPBridge commented:
@fedorov reminder about this

fedorov committed:
refinements to direct access page

* added section on getting file URLs using idc-index
* updated all examples with the details on getting bucket file URLs using idc-index
* replaced direct use of tag numbers with keyword lookup
* updated importance of offset tables to clarify it is about SM

This notebook contains all of the code samples for convenient testing - I plan to
update this notebook and add it to IDC-Tutorials.

https://colab.research.google.com/drive/1_5cc_kb-FgUl9r0jePX_QbSvwBFEluMg?usp=sharing

fedorov commented:
@CPBridge I updated, please review and let me know if you have any concerns!

CPBridge committed:
Tweaks to index examples

CPBridge committed:
Minor tweaks

CPBridge commented:
Thanks @fedorov. I found a few bugs in your examples (some lines looked like they were out of date or spliced from the wrong example). So I fixed them and made some minor cosmetic changes at the same time.

One small concern I have is while the method used to find the largest image in the pyramid nicely demonstrates the some of the capabilities listed on the page, it is also quite inefficient and not how I would recommend people do this. I would recommend doing this either with bigquery or even idc-index, which if memory serves has an instance-level table for SM images that would contain this information. What do you think? I think at the very least we should point out the availability of better options

CPBridge commented:
P.s. I also added a small snippet to figure out whether a given image has an EOT/BOT, which I figured out a few weeks ago for something else but it might be useful on this page

fedorov committed:
replace iteration over instances with direct selection using new idc-index function

fedorov commented:
> I found a few bugs in your examples

Thank you for fixing them, sorry!

> I would recommend doing this either with bigquery or even idc-index, which if memory serves has an instance-level table for SM images that would contain this information. What do you think?

It took me a little bit more than I hoped, but I added a helper function (otherwise it would be ugly to build AWS/GCP-specific URLs) in idc-index 0.8.7, and update the relevant code snippet!

If this looks good, I am going to merge - let me know!

CPBridge commented:
@fedorov looks good to me!
